core/trino-main/src/main/java/io/trino/simd/BlockEncodingSimdSupport.java

```diff
@@ -0,0 +1,118 @@
+
+/*
+We need to specifically detect AVX512F (for VPCOMPRESSD / VPCOMPRESSQ instruction support for int and long types) and
+AVX512VBMI2 (for VPCOMPRESSB / VPCOMPRESSW instruction support over byte and short types). Because we would like to check
+whether Vector<T>#compress(VectorMask<T>) is supported natively in hardware or emulated by the JVM - because the emulated
+support is so much slower than the simple scalar code that exists, but since we don't have the ability to detect that
+directly from the JDK vector API we have to assume that native support exists whenever the CPU advertises it.
+ */
+@Singleton
+public final class BlockEncodingSimdSupport
+{
+    public record SimdSupport(
+            boolean expandAndCompressByte,
+            boolean expandAndCompressShort,
+            boolean expandAndCompressInt,
+            boolean expandAndCompressLong)
+    {
+        public static final SimdSupport NONE = new SimdSupport(false, false, false, false);
+        public static final SimdSupport ALL = new SimdSupport(true, true, true, true);
+    }
+
+    public static final int MINIMUM_SIMD_LENGTH = 512;
+    private final SimdSupport simdSupport;
+    private static final SimdSupport AUTO_DETECTED_SUPPORT = detectSimd();
+
+    public static final BlockEncodingSimdSupport TESTING_BLOCK_ENCODING_SIMD_SUPPORT = new BlockEncodingSimdSupport(true);
+
+    @Inject
+    public BlockEncodingSimdSupport(
+            FeaturesConfig featuresConfig)
+    {
+        this(featuresConfig.isBlockSerdeVectorizedNullSuppressionEnabled());
+    }
+
+    public BlockEncodingSimdSupport(
+            boolean enableAutoDetectedSimdSupport)
+    {
+        if (enableAutoDetectedSimdSupport) {
+            simdSupport = AUTO_DETECTED_SUPPORT;
+        }
+        else {
+            simdSupport = SimdSupport.NONE;
+        }
+    }
+
+    private static SimdSupport detectSimd()
+    {
+        ProcessorIdentifier id = MachineInfo.getProcessorInfo();
+
+        String vendor = id.getVendor().toLowerCase(ENGLISH);
+
+        if (vendor.contains("intel") || vendor.contains("amd")) {
+            return detectX86SimdSupport();
+        }
+
+        return SimdSupport.NONE;
+    }
+
+    private static SimdSupport detectX86SimdSupport()
+    {
+        enum X86SimdInstructionSet {
+            avx512f,
+            avx512vbmi2
+        }
+
+        Set<String> flags = readCpuFlags();
+        EnumSet<X86SimdInstructionSet> x86Flags = EnumSet.noneOf(X86SimdInstructionSet.class);
+
+        if (!flags.isEmpty()) {
+            for (X86SimdInstructionSet instructionSet : X86SimdInstructionSet.values()) {
+                if (flags.contains(instructionSet.name())) {
+                    x86Flags.add(instructionSet);
+                }
+            }
+        }
+
+        return new SimdSupport(
+                (ByteVector.SPECIES_PREFERRED.vectorBitSize() >= MINIMUM_SIMD_LENGTH) && x86Flags.contains(X86SimdInstructionSet.avx512vbmi2),
```

Comment:
Why is the calculation `ByteVector.SPECIES_PREFERRED.vectorBitSize()` per Vector type ?
Don't they all return the same value as `VectorShape.preferredShape().vectorBitSize()` ?
I assume we're looking for bit width of SIMD register on the CPU here, so that should be available from one method call.